lib/cache/collections.go

```diff
@@ -2373,6 +2381,56 @@ func (accessListsExecutor) getReader(cache *Cache, cacheOK bool) services.Access
 
 var _ executor[*accesslist.AccessList, services.AccessListsGetter] = accessListsExecutor{}
 
+type discoveryConfigExecutor struct{}
+
+func (discoveryConfigExecutor) getAll(ctx context.Context, cache *Cache, loadSecrets bool) ([]*discoveryconfig.DiscoveryConfig, error) {
+	var discoveryConfigs []*discoveryconfig.DiscoveryConfig
+	var nextToken string
+	for {
+		var page []*discoveryconfig.DiscoveryConfig
+		var err error
+
+		page, nextToken, err = cache.DiscoveryConfigs.ListDiscoveryConfigs(ctx, 0 /* default page size */, nextToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		discoveryConfigs = append(discoveryConfigs, page...)
+
+		if nextToken == "" {
+			break
+		}
+	}
+	return discoveryConfigs, nil
+}
+
+func (discoveryConfigExecutor) upsert(ctx context.Context, cache *Cache, resource *discoveryconfig.DiscoveryConfig) error {
+	_, err := cache.discoveryConfigsCache.CreateDiscoveryConfig(ctx, resource)
+	if trace.IsAlreadyExists(err) {
+		_, err = cache.discoveryConfigsCache.UpdateDiscoveryConfig(ctx, resource)
+	}
```

Comment:
This is a terrible approach to upsert, and it's not really justifiable for a brand new resource type IMO. Can we add a real upsert operation, outright drop the implementation of `UpdateDiscoveryConfig` in v14 and only support it in v15+ as a conditional update operation instead?

Reply:
I've just added the upsert method.

> outright drop the implementation of UpdateDiscoveryConfig in v14 and only support it in v15+ as a conditional update operation instead?

Sorry, not sure I followed.
Are you suggesting to remove the UpdateDiscoveryConfig in v15/master but keep it in v14?
But then clients using v14 will hit the UpdateDiscoveryConfig and it will fail

Could you please clear this up?

(maybe by v15+ you meant v16 and ownards, not sure if your `+` was inclusive)


This resource is not yet in use, but we should be careful anyway because it was backported to v14

Reply:
Since v14.0.1 doesn't have an implementation for `UpdateDiscoveryConfig` yet, I propose that we remove `UpdateDiscoveryConfig` from branch/v14 and leave it as a not implemented RPC, add (and backport) `UpsertDiscoveryConfig` from this PR and use it in v14 clients, then in master and branch/v15 we can implement it as a conditional update operation.

Reply:
> remove UpdateDiscoveryConfig from branch/v14 and leave it as a not implemented RPC

It might get released before we are able to merge the backport.
I think, the Update method is the intended one for 99% of the cases.
That also holds true when we implement the condition updates/revisions.

Reply:
What is "write-if-any-exists" useful for?

Reply:
Well, honestly the only other client (for now) is going to be the WebAPI
I'm not sure we need it there.

This should ease the path for the revision logic we intend to implement later.

I would rather not remove it.